### PR TITLE
Add valid data (+TVN fixes)

### PR DESCRIPTION
**Pull request by TevenLeScao**
Re-creating the pull request at #113 so I can push directly

### CODE REVIEW

**Comment by TevenLeScao** [comment]
I've taken over the PR from @sbmaruf to change the behaviour of the validation argument when the same dataset is passed to train and to validation. In this case, I re-use the pre-existing split in train and validation that was created previously, instead of putting 100% of the dataset in a validation split (which would create overlap between train and valid).

File: megatron/data/gpt_dataset.py [hunk 1]
```diff
@@ -48,27 +50,61 @@ def build_train_valid_test_datasets(data_prefix, data_impl, splits_string,
...
+        print(f"split: {splits_string}")
         if train_ds:
             train_datasets.append(train_ds)
+            print(f"train_ds size: {len(train_ds)}")
         if valid_ds:
-            valid_datasets.append(valid_ds)
+            valid_datasets_dict[prefix] = valid_ds
+            print(f"valid_ds size: {len(valid_ds)}")
         if test_ds:
             test_datasets.append(test_ds)
```
Comment (stas00) [inline megatron/data/gpt_dataset.py:71]: please careful with adding direct prints as some of these will be replicated 512 times or more. You want `print_rank_0` instead.

Also should the test_ds branch have one for consistency?

Comment (stas00) [reply megatron/data/gpt_dataset.py:71]: same in the rest of this PR

File: megatron/data/gpt_dataset.py [hunk 2]
```diff
@@ -48,27 +50,61 @@ def build_train_valid_test_datasets(data_prefix, data_impl, splits_string,
...
+        # in this case we only keep the validation sets in the arguments
+        valid_output = get_datasets_weights_and_num_samples(valid_data_prefix,
+                                                  [0, train_valid_test_num_samples[1], 0])
+        valid_prefixes, valid_weights, valid_datasets_samples = valid_output
+        for i, prefix in enumerate(valid_prefixes):
+            if prefix not in valid_datasets_dict:
+                print(f"prefix: {prefix} not found in {valid_datasets_dict.keys()}")
+                # create the ones from the arguments that are missing
+                train_ds, valid_ds, test_ds = _build_train_valid_test_datasets(
+                    valid_prefixes[i], data_impl, '0,100,0',
```
Comment (stas00) [inline megatron/data/gpt_dataset.py:83]: should the presence of this new flag require the user to pass the normal data-path split so that the 2nd split is 0, e.g. `100,0,0` or `90,0,10`?

Will this make things simpler?

File: megatron/arguments.py [hunk 3]
```diff
@@ -648,6 +648,11 @@ def _add_data_args(parser):
                        '1) a single data path, 2) multiple datasets in the'
                        'form: dataset1-weight dataset1-path dataset2-weight '
                        'dataset2-path ...')
+    group.add_argument('--valid-data', nargs='*', default=None,
```
Comment (stas00) [inline megatron/arguments.py:651]:
```suggestion
    group.add_argument('--valid-data-path', nargs='*', default=None,
```

Please, let's use consistent naming. And if accepted there will be a need for a corresponding change later in `args.valid_data`.

File: megatron/data/gpt_dataset.py (hunk 1 above)
Comment (TevenLeScao) [reply megatron/data/gpt_dataset.py]: Changed the prints ! Half thinking we should replace every single print in the repo by print_rank_0 now

File: megatron/data/gpt_dataset.py (hunk 2 above)
Comment (stas00) [reply megatron/data/gpt_dataset.py:83]: Or I'd go even simpler and require no split argument then, if a separate validation data path is passed then do no splitting at all. i.e. data-path => Train, valid-data-path => Valid (and we can add test-data-path if needed).

Yet another possible solution:

Leave -data-path feature as is, 

And have `--(train|valid|test)-data-path` arg and make these mutually exclusive.

So it's either:
1. `--data-path` and `--split` pair or
2. `--(train|valid|test)-data-path` (3 args) and no `--split` 

The originally proposed solution feels too much of a "patch".

File: megatron/data/gpt_dataset.py (hunk 1 above)
Comment (TevenLeScao) [reply megatron/data/gpt_dataset.py]: I was on the fence on the test_ds - I didn't want to add more code for something that we don't currently use. I'll try to think of a way to factor it

File: megatron/arguments.py (hunk 3 above)
Comment (TevenLeScao) [reply megatron/arguments.py:651]: Ah yes for sure, missed this

File: megatron/data/gpt_dataset.py (hunk 2 above)
Comment (TevenLeScao) [reply megatron/data/gpt_dataset.py:83]: Something I feel we need with this feature is to deal with datasets that only have a `train` split and no `validation`, including OSCAR which is why I wrote it like this. In this case you still want to be able to act on your validation mix even though you don't have a separate file to point to: for example, you want to use only one of the validation splits you created in the preceding step rather the mix of all.

Comment (TevenLeScao) [reply megatron/data/gpt_dataset.py:83]: If you point to a file that's not already in the training set, e.g. an external validation set, then it feels natural to use it all. We can easily add another flag to split it but I cannot see a usecase for it.

Comment (stas00) [reply megatron/data/gpt_dataset.py:83]: If I understand your correctly that's why I suggested to have 2 different sets of APIs:

1. `--data-path` and `--split` pair - normal use
2. `--(train|valid|test)-data-path` (3 args) and no `--split` and test is optional. For when you want to take full control over the splits. 

I think this would make things easier to use.

but perhaps I'm not seeing some nuance here.

**Comment by stas00** [comment]
re: testing - we have 2 types of tests:
1. does it run
2. quality tests

the first type of test is trivial, just clone one of the existing main training tests (test_training.py) and add the new options, done (we will then refactor repetition as tests add up).

the second group is think of how to validate this, and you can also leave it as a separate Issue and perhaps someone will contribute the missing part of the test.

but at the very least let's make sure the framework runs with the new features.

**Comment by TevenLeScao** [comment]
Looks like I was too careless with my replacing of print_rank_0s, the rest is codecarbon!

File: megatron/data/gpt_dataset.py (hunk 2 above)
Comment (stas00) [reply megatron/data/gpt_dataset.py:83]: See, wearing a user hat, this is very confusing to me. 

You propose to have 2 different ways to define a validation dataset, and how they work together is very unclear to me.

At the very least I'd then call it `--extra-valid-data-path` so it's clear that it's an additional validation dataset.

Otherwise the interface is very confusing if you include validations from 2 arguments that behave differently - the original one originates from a split arg, the 2nd one doesn't.

Comment (stas00) [reply megatron/data/gpt_dataset.py:83]: And that's why I proposed to not overlap the 2, and instead define a single interface to adding a validation dataset, rather than a confusing overlap.

Comment (TevenLeScao) [reply megatron/data/gpt_dataset.py:83]: It is not an extra valid data path:  when passed, the argument defines on its own the integrality of validation source files, and makes sure there is no overlap with training data. I feel like that part's naming is quite transparent to the user: if you pass it, the validation data comes exactly from the files you defined. There's no part A and part B, just what you defined under that flag. The part that may not be clear is that it also prevents overlap/respects the splits you passed to `data-path`.

Comment (TevenLeScao) [reply megatron/data/gpt_dataset.py:83]: > A user will think that if -valid-data-path is declared, this is their only source of validation data - why would it be any different.

That is exactly what happens!

Comment (stas00) [reply megatron/data/gpt_dataset.py:83]: I'm pretty sure this is not what your code is doing ;) if you follow closely the if/else branches. Please revisit lines 78-91.

The only thing I can see clearly is that you ensure that the same dataset is not listing twice.

Comment (TevenLeScao) [reply megatron/data/gpt_dataset.py:84]: What I'm reading is that if `valid_data_path` is not `None`, every single prefix that gets added to `valid_datasets` originally comes from `get_datasets_weights_and_num_samples(valid_data_prefix, [0, train_valid_test_num_samples[1], 0])`

Comment (TevenLeScao) [reply megatron/data/gpt_dataset.py:84]: So it seems to me that `valid_data_path` is the variable that defines what gets added or not: all of the datasets that get tested for overlap and eventually added originally come from it.

Comment (TevenLeScao) [reply megatron/data/gpt_dataset.py:84]: Ah, a comment was misleading because it was wrongly copy-pasted, I fixed it!

Comment (TevenLeScao) [reply megatron/data/gpt_dataset.py:85]: I also want it to only produce b, and I think this is what happens.

File: megatron/data/gpt_dataset.py [hunk 4]
```diff
@@ -48,27 +50,63 @@ def build_train_valid_test_datasets(data_prefix, data_impl, splits_string,
...
 
+    if valid_data_prefix is not None:
+        # in this case `valid_data_prefix` defines what is in the validation mix
+        # we make sure there is no overlap with the training mix by comparing with ds_shared_with_train
+        valid_output = get_datasets_weights_and_num_samples(valid_data_prefix,
+                                                  [0, train_valid_test_num_samples[1], 0])
+        valid_prefixes, valid_weights, valid_datasets_samples = valid_output
+        for i, prefix in enumerate(valid_prefixes):
+            if prefix not in ds_shared_with_train:
+                print_rank_0(f"prefix: {prefix} not found in {ds_shared_with_train.keys()}")
```
Comment (stas00) [inline megatron/data/gpt_dataset.py:82]:
```suggestion
                print_rank_0(f"prefix: {prefix} not found in {ds_shared_with_train.keys()}")
```
This looks more like a debug and won't be of any added value to the user, IMHO, and just add to the noise.

What do you think?

File: megatron/data/gpt_dataset.py (hunk 2 above)
Comment (stas00) [reply megatron/data/gpt_dataset.py:85]: Thank you for bearing with me, Teven, I can see what you mean now wrt to the last part. I was missing that the prefixes were coming only from the new variable.

So you will be still getting a mix of 2 ways if the user were to call
```
--data-path a --split 80,20,0
--valid-data-path a 10 b 20
```
as you will get a slice of `a` and full `b`.

I think it'd be much safer to assert if the same prefix is used in both. It clearly is a wrong use.

Comment (stas00) [reply megatron/data/gpt_dataset.py:85]: moreover in the last example I have shown, which weights apply? The weights coming from `--data-path` or from `--valid-data-path` - I think you're now mixing the two.

Comment (TevenLeScao) [reply megatron/data/gpt_dataset.py:85]: It wasn't bearing, the code needed to be clearer anyway. The weights of valid-data-path apply, which is clear since we're also using the data from valid-data-path!

I am not sure what you call a wrong use - in the OSCAR multilingual experiments, we 1. do not have a separate validation set 2. we need to restrict the validation to only a language or set of languages. This naturally leads to passing a path that overrides the original validation mix but also has overlap with it.

Comment (sbmaruf) [reply megatron/data/gpt_dataset.py:85]: I agree with @stas00  

Why are you selecting validation data from training data? You can just create data in `--valid-path` and take it from there. To me it seems redundant. 

The implementation was, 

You only take `validation` data from training when you don't have any data path `--valid-data-path` argument. When you are taking data from both `--data-path` and `--valid-data-path`, that's so confusing to me. 

I understand this is needed for OSCAR. But I am against it because in this way, we cannot track which samples are  selected for training and which one for development. IMO if oscar doesn't have development set, we should create our own split and share that split for re-producibility. 

Another problem, 

Let's assume a case, 

```
TRAIN_DATASET=DATASET_0
VALID_DATASET="0.1 ${DATASET_0} 0.25 ${DATASET_1} 0.2 ${DATASET_2} 0.15 ${DATASET_3} 0.3 ${DATASET_4}"
```

when you run this with `--split 80,20,0`, 20% of training data with be sampled with a relative probability `0.1`. For a normal user, this seems overly complicated.

Comment (TevenLeScao) [reply megatron/data/gpt_dataset.py:85]: Yes, this is different from your implementation - as discussed above, this only takes data from `--valid-data-path` and **not from --data-path**. We can talk about the need for a separate development set for OSCAR but it is not the only dataset in this case as language modeling generally doesn't need validation sets. I'd rather give the user (i.e. us!) this flexilibity rather than forcing them to make their own splits.

Comment (stas00) [reply megatron/data/gpt_dataset.py:85]: I'm glad to hear that it's not just me who finds the proposed API confusing and the fact that were are still trying to figure it out is an indication that.

May I re-iterate a different proposal, I suggested yesterday, 

where we have two mutually exclusive modes:

1. `--data-path` and `--split` pair - normal use
2. `--(train|valid|test)-data-path` (3 args) and no `--split` and test is optional. For when you want to take full control over the splits. 

I think this would make things easier to use, since the 2nd approach is very explicit.
